omniledger/service/service.go

```diff
@@ -280,11 +280,36 @@ func (s *Service) GetProof(req *GetProof) (resp *GetProofResponse, err error) {
 		return nil, errors.New("version mismatch")
 	}
 	log.Lvlf2("%s Getting proof for key %x on sc %x", s.ServerIdentity(), req.Key, req.ID)
-	latest, err := s.db().GetLatestByID(req.ID)
-	if err != nil && latest == nil {
+
+	// Get the collection database and the latest skipblock, if the latest
+	// skipblock does not have the same index as the one in the collection
+	// then we look back for one that does.
+	sb, err := s.db().GetLatestByID(req.ID)
+	if err != nil && sb == nil {
+		return
+	}
+	cdb := s.getCollection(req.ID)
+	if cdb.getIndex() > sb.Index {
+		err = errors.New("invalid index while getting proof")
+		return
+	}
+	// TODO can cdb.getIndex change in this loop?
```

Comment:
Yes, it most certainly can. Probably the right answer is that NewProof has to return the index number it was working from when it created the proof. But that's not so easy, because the locking for the read/write of the index number is managed by boltdb, and the locking for the duration of the creation of the proof happens inside of c.coll.Get(key).Proof() in NewProof().

Ugh.

It is very much bad and wrong that we have a memory and a disk copy of one thing and that they are not accessed via one unified API which can manage the locking of them as an all or nothing unit.

Reply:
Yeah, I'm not too sure what we can do about it.. One thing is we store the first index and make sure it doesn't change during the looping, that will at least avoid inconsistency but clients will have to try again.